# Data Cleaning

We are working with the Online Retail II dataset which can be found here: https://archive.ics.uci.edu/dataset/502/online+retail+ii.

**Given Dataset Description:**  

**InvoiceNo:** Invoice number. Nominal. A 6-digit integral number uniquely assigned to each transaction. If this code starts with the letter 'c', it indicates a cancellation.  
**StockCode:** Product (item) code. Nominal. A 5-digit integral number uniquely assigned to each distinct product.  
**Description:** Product (item) name. Nominal.  
**Quantity:** The quantities of each product (item) per transaction. Numeric.  
**InvoiceDate:** Invice date and time. Numeric. The day and time when a transaction was generated.  
**UnitPrice:** Unit price. Numeric. Product price per unit in sterling (£).  
**CustomerID:** Customer number. Nominal. A 5-digit integral number uniquely assigned to each customer.  
**Country:** Country name. Nominal. The name of the country where a customer resides.  

In [1]:
import numpy as np
import pandas as pd

In [2]:
# Load sheets into separate dataframes
df1 = pd.read_excel('data/online_retail_II.xlsx', sheet_name='Year 2009-2010')
df2 = pd.read_excel('data/online_retail_II.xlsx', sheet_name='Year 2010-2011')

# Stack the rows together and reset the index
df = pd.concat([df1, df2], ignore_index=True)

In [3]:
df.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom


In [4]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1067371 entries, 0 to 1067370
Data columns (total 8 columns):
 #   Column       Non-Null Count    Dtype         
---  ------       --------------    -----         
 0   Invoice      1067371 non-null  object        
 1   StockCode    1067371 non-null  object        
 2   Description  1062989 non-null  object        
 3   Quantity     1067371 non-null  int64         
 4   InvoiceDate  1067371 non-null  datetime64[us]
 5   Price        1067371 non-null  float64       
 6   Customer ID  824364 non-null   float64       
 7   Country      1067371 non-null  str           
dtypes: datetime64[us](1), float64(2), int64(1), object(3), str(1)
memory usage: 78.8+ MB


### Rename Columns

In [5]:
df = df.rename(columns={'Customer ID': 'CustomerID'})
df.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,CustomerID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom


### Check Datatypes

In [6]:
# Check for mixed types within each object column
for col in df.columns:
    types_present = df[col].dropna().apply(type).value_counts()
    print(col, '->', types_present.to_dict())

Invoice -> {<class 'int'>: 1047871, <class 'str'>: 19500}
StockCode -> {<class 'int'>: 932385, <class 'str'>: 134986}
Description -> {<class 'str'>: 1062985, <class 'int'>: 4}
Quantity -> {<class 'int'>: 1067371}
InvoiceDate -> {<class 'pandas.Timestamp'>: 1067371}
Price -> {<class 'float'>: 1067371}
CustomerID -> {<class 'float'>: 824364}
Country -> {<class 'str'>: 1067371}


In [7]:
df['Invoice'] = df['Invoice'].astype(str)
df['StockCode'] = df['StockCode'].astype(str)
df['CustomerID'] = df['CustomerID'].astype('Int64')  # nullable int, keeps NaN as <NA>, no more .0
df['Description'] = df['Description'].astype('str')

### Drop Duplicate Rows

In [8]:
print(f"Duplicate Rows: {df.duplicated().sum()}")
df.drop_duplicates(inplace=True)
print(f"Length of the dataset after dropping duplicates: {len(df)}")

Duplicate Rows: 34335
Length of the dataset after dropping duplicates: 1033036


In [9]:
df.info()

<class 'pandas.DataFrame'>
Index: 1033036 entries, 0 to 1067370
Data columns (total 8 columns):
 #   Column       Non-Null Count    Dtype         
---  ------       --------------    -----         
 0   Invoice      1033036 non-null  str           
 1   StockCode    1033036 non-null  str           
 2   Description  1028761 non-null  str           
 3   Quantity     1033036 non-null  int64         
 4   InvoiceDate  1033036 non-null  datetime64[us]
 5   Price        1033036 non-null  float64       
 6   CustomerID   797885 non-null   Int64         
 7   Country      1033036 non-null  str           
dtypes: Int64(1), datetime64[us](1), float64(1), int64(1), str(4)
memory usage: 122.7 MB


### Handle the Datetime Column

In [10]:
from datetime import datetime

df['InvoiceDateTime'] = df['InvoiceDate'] 
df['InvoiceDate'] = df['InvoiceDateTime'].dt.date
df['InvoiceTime'] = df['InvoiceDateTime'].dt.time
df['InvoiceHour'] = df['InvoiceDateTime'].dt.hour
df['InvoiceDayOfWeek'] = df['InvoiceDateTime'].dt.day_name()
df['InvoiceMonth'] = df['InvoiceDateTime'].dt.to_period('M')

In [11]:
df.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,CustomerID,Country,InvoiceDateTime,InvoiceTime,InvoiceHour,InvoiceDayOfWeek,InvoiceMonth
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01,6.95,13085,United Kingdom,2009-12-01 07:45:00,07:45:00,7,Tuesday,2009-12
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01,6.75,13085,United Kingdom,2009-12-01 07:45:00,07:45:00,7,Tuesday,2009-12
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01,6.75,13085,United Kingdom,2009-12-01 07:45:00,07:45:00,7,Tuesday,2009-12
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01,2.10,13085,United Kingdom,2009-12-01 07:45:00,07:45:00,7,Tuesday,2009-12
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01,1.25,13085,United Kingdom,2009-12-01 07:45:00,07:45:00,7,Tuesday,2009-12


### Handle Missing Values

**CustomerID**

In [12]:
df.isna().sum()

Invoice                  0
StockCode                0
Description           4275
Quantity                 0
InvoiceDate              0
Price                    0
CustomerID          235151
Country                  0
InvoiceDateTime          0
InvoiceTime              0
InvoiceHour              0
InvoiceDayOfWeek         0
InvoiceMonth             0
dtype: int64

Customer ID is an important field for customer-level analysis. Hence, we can check:
- How much revenue/volume is at stake - does excluding them lose a trivial or meaningful chunk?
- Do guest orders tend to be smaller on average than identified-customer orders?

In [13]:
missing_rows = df[df['CustomerID'].isna()]
print(f"Missing-ID row share: {len(missing_rows)/len(df):.2%}")
print(f"Missing-ID revenue share: {(missing_rows['Quantity']*missing_rows['Price']).sum() / (df['Quantity']*df['Price']).sum():.2%}")

Missing-ID row share: 22.76%
Missing-ID revenue share: 13.61%


In [14]:
guest_avg_order = missing_rows.groupby('Invoice').apply(lambda x: (x['Quantity']*x['Price']).sum()).mean()
identified_avg_order = df[~df['CustomerID'].isna()].groupby('Invoice').apply(lambda x: (x['Quantity']*x['Price']).sum()).mean()
print(f"Guest avg order value: £{guest_avg_order:.2f}")
print(f"Identified avg order value: £{identified_avg_order:.2f}")

Guest avg order value: £293.14
Identified avg order value: £363.00


- 22.76% seems like a sizeable amount of missing customer IDs to just throw away. We can still use them in product-level or basket-level analysis.
- 13.6% of revenue share shows a good amount of revenue comes from unidentified customers.
- Also, the guest average order value is not very far from the identified customers' average order value.
- It would benefit to keep these transactions with missing customer ID to analyse other things.
- Since Invoice is generated for one purchase, we can use it generate the customer ID for some guests. The transactions with the same invoice can be grouped, as they would be from the same customer.
- Here's how we'll handle this: We'll create a unique pseudo ID = 'GUEST' + invoice number. Also, we'll create a binary flag 'is_guest'.
- This will help us identify guests and also handle the missing values.

In [15]:
df['CustomerID'] = df['CustomerID'].astype('object')
missing_mask = df['CustomerID'].isna()

df['is_guest'] = missing_mask
df.loc[missing_mask, 'CustomerID'] = 'GUEST_' + df.loc[missing_mask, 'Invoice'].astype(str)

In [16]:
# Verify
df[df['CustomerID'].astype(str).str.startswith('GUEST_')].head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,CustomerID,Country,InvoiceDateTime,InvoiceTime,InvoiceHour,InvoiceDayOfWeek,InvoiceMonth,is_guest
263,489464,21733,85123a mixed,-96,2009-12-01,0.00,GUEST_489464,United Kingdom,2009-12-01 10:52:00,10:52:00,10,Tuesday,2009-12,True
283,489463,71477,short,-240,2009-12-01,0.00,GUEST_489463,United Kingdom,2009-12-01 10:52:00,10:52:00,10,Tuesday,2009-12,True
284,489467,85123A,21733 mixed,-192,2009-12-01,0.00,GUEST_489467,United Kingdom,2009-12-01 10:53:00,10:53:00,10,Tuesday,2009-12,True
470,489521,21646,NaN,-50,2009-12-01,0.00,GUEST_489521,United Kingdom,2009-12-01 11:44:00,11:44:00,11,Tuesday,2009-12,True
577,489525,85226C,BLUE PULL BACK RACING CAR,1,2009-12-01,0.55,GUEST_489525,United Kingdom,2009-12-01 11:49:00,11:49:00,11,Tuesday,2009-12,True


In [17]:
# Verify
df['CustomerID'].isna().sum()

np.int64(0)

**Description**

In case of the Description column, in case time permits, we can implement a NLP customer segmentation to identify the different products purchased and analyse the product-level stats.  

In [18]:
df['Description'].isna().sum()

np.int64(4275)

Let's check if every StockCode has a unique description. If that's the case, we can use descriptions from other transactions to fill the misisng values.

In [19]:
# How many distinct descriptions does each StockCode have?
desc_per_code = df.groupby('StockCode')['Description'].nunique()
print(desc_per_code.value_counts().sort_index())

# StockCodes with more than 1 description — worth inspecting
inconsistent_codes = desc_per_code[desc_per_code > 1].index
print(f"StockCodes with multiple descriptions: {len(inconsistent_codes)}")
df[df['StockCode'].isin(inconsistent_codes)].sort_values('StockCode')[['StockCode','Description']].drop_duplicates().head(20)

Description
0     355
1    3718
2     953
3     204
4      57
5       9
6       4
7       4
9       1
Name: count, dtype: int64
StockCodes with multiple descriptions: 1232


,StockCode,Description
6339,10080,GROOVY CACTUS INFLATABLE
183163,10080,NaN
979826,10080,check
2944,10120,DOGGY RUBBER
194372,10120,Zebra invcing error
475,10133,COLOURING PENCILS BROWN TUBE
1000108,10133,damaged
614,15056N,EDWARDIAN PARASOL NATURAL
100429,15056N,wedding co returns?
67142,15058A,BLUE WHITE SPOTS GARDEN PARASOL


- There are 1232 stock codes with multiple descriptions.
- Notice that every legitimate product description is in ALL CAPS, while there are certain junk entries in lowercase or mixed case.
- These look like internal stock-adjustment notes typed by warehouse staff into the same free-text field, not real product descriptions. They're probably associated with adjustment/write-off transactions (damaged stock, mis-shipments, stocktake corrections) rather than genuine customer sales.
- We will replace both, the missing descriptions (NA) and the junk descriptions with the most-used descriptions from the dataset(modes).

In [20]:
# Canonical description = most frequent Description for each StockCode
# Note that we have only considered valid transactions where the Price is positive and not 0
canonical_desc = df[df['Price'] > 0].groupby('StockCode')['Description'].agg(lambda x: x.mode().iloc[0] if not x.mode().empty else np.nan)

# Missing descriptions will be filled with the description that is most used for that stockCode
# Same for junk values (values that are ued less times)
df['Description'] = df['StockCode'].map(canonical_desc)

In [21]:
# Check remaining missing values
remaining_missing = df['Description'].isna().sum()
print(f"Remaining missing descriptions: {remaining_missing}")
df[df['Description'].isna()]

Remaining missing descriptions: 393


,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,CustomerID,Country,InvoiceDateTime,InvoiceTime,InvoiceHour,InvoiceDayOfWeek,InvoiceMonth,is_guest
470,489521,21646,NaN,-50,2009-12-01,0.0,GUEST_489521,United Kingdom,2009-12-01 11:44:00,11:44:00,11,Tuesday,2009-12,True
3162,489660,35956,NaN,-1043,2009-12-01,0.0,GUEST_489660,United Kingdom,2009-12-01 17:43:00,17:43:00,17,Tuesday,2009-12,True
3168,489663,35605A,NaN,-117,2009-12-01,0.0,GUEST_489663,United Kingdom,2009-12-01 18:02:00,18:02:00,18,Tuesday,2009-12,True
16186,490736,35983,NaN,-200,2009-12-07,0.0,GUEST_490736,United Kingdom,2009-12-07 17:13:00,17:13:00,17,Monday,2009-12,True
19313,490976,84571,NaN,-120,2009-12-08,0.0,GUEST_490976,United Kingdom,2009-12-08 16:42:00,16:42:00,16,Tuesday,2009-12,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
804718,561254,DCGS0067,NaN,-11,2011-07-26,0.0,GUEST_561254,United Kingdom,2011-07-26 11:52:00,11:52:00,11,Tuesday,2011-07,True
872310,567207,35592T,NaN,4,2011-09-19,0.0,GUEST_567207,United Kingdom,2011-09-19 11:01:00,11:01:00,11,Monday,2011-09,True
942568,572614,23595,NaN,5,2011-10-25,0.0,GUEST_572614,United Kingdom,2011-10-25 11:38:00,11:38:00,11,Tuesday,2011-10,True
942569,572615,23595,NaN,-5,2011-10-25,0.0,GUEST_572615,United Kingdom,2011-10-25 11:39:00,11:39:00,11,Tuesday,2011-10,True


In [22]:
df[df['Description'].isna()]['CustomerID'].str.startswith('GUEST').sum()

np.int64(393)

In [23]:
df[df['Description'].isna()]['Price'].value_counts()

Price
0.0    393
Name: count, dtype: int64

In [24]:
df[df['Description'].isna()]['Quantity'].value_counts()

Quantity
-2       35
-1       35
 1       20
-3       18
-4       14
         ..
-6400     1
-524      1
-2472     1
-2376     1
 200      1
Name: count, Length: 118, dtype: int64

- There are still 393 rows with missing descriptions. This means these StockCodes do not have any value for any of the transactions done. 
- Note that the Customer ID for all of them was missing in the dataset (GUEST customers) and also the Price is 0 for all these entries.
- These are probably manual entries or fixes done by the staff.
- As the price is 0, they will not be helpful for revenue analysis and with no description, we cannot identify the products either.
- We can safely drop all these rows from our analysis. 

In [25]:
# These 393 rows carry no real product identity or price — treat as noise,
# consistent with the stock-adjustment/non-product handling done earlier
df = df[~df['Description'].isna()]

In [26]:
# Verify
df['Description'].isna().sum()

np.int64(0)

### Analyse each column to identify any issues

In [27]:
df.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,CustomerID,Country,InvoiceDateTime,InvoiceTime,InvoiceHour,InvoiceDayOfWeek,InvoiceMonth,is_guest
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01,6.95,13085,United Kingdom,2009-12-01 07:45:00,07:45:00,7,Tuesday,2009-12,False
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01,6.75,13085,United Kingdom,2009-12-01 07:45:00,07:45:00,7,Tuesday,2009-12,False
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01,6.75,13085,United Kingdom,2009-12-01 07:45:00,07:45:00,7,Tuesday,2009-12,False
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01,2.10,13085,United Kingdom,2009-12-01 07:45:00,07:45:00,7,Tuesday,2009-12,False
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01,1.25,13085,United Kingdom,2009-12-01 07:45:00,07:45:00,7,Tuesday,2009-12,False


In [28]:
# Extract any leading non-digit characters
invoice_prefix = df['Invoice'].str.extract(r'^([^\d]*)')[0]
print(invoice_prefix.value_counts())

0
     1013533
C      19104
A          6
Name: count, dtype: int64


In [29]:
df[df['Invoice'].str.startswith('A')]

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,CustomerID,Country,InvoiceDateTime,InvoiceTime,InvoiceHour,InvoiceDayOfWeek,InvoiceMonth,is_guest
179403,A506401,B,Adjust bad debt,1,2010-04-29,-53594.36,GUEST_A506401,United Kingdom,2010-04-29 13:36:00,13:36:00,13,Thursday,2010-04,True
276274,A516228,B,Adjust bad debt,1,2010-07-19,-44031.79,GUEST_A516228,United Kingdom,2010-07-19 11:24:00,11:24:00,11,Monday,2010-07,True
403472,A528059,B,Adjust bad debt,1,2010-10-20,-38925.87,GUEST_A528059,United Kingdom,2010-10-20 12:04:00,12:04:00,12,Wednesday,2010-10,True
825443,A563185,B,Adjust bad debt,1,2011-08-12,11062.06,GUEST_A563185,United Kingdom,2011-08-12 14:50:00,14:50:00,14,Friday,2011-08,True
825444,A563186,B,Adjust bad debt,1,2011-08-12,-11062.06,GUEST_A563186,United Kingdom,2011-08-12 14:51:00,14:51:00,14,Friday,2011-08,True
825445,A563187,B,Adjust bad debt,1,2011-08-12,-11062.06,GUEST_A563187,United Kingdom,2011-08-12 14:52:00,14:52:00,14,Friday,2011-08,True


In [30]:
df = df[~df['Invoice'].str.startswith('A')]

- There are only 6 entries with Invoice numbers starting with 'A'. 
- Note that the stockCode is B for them all, quantity is 1 and the price is negative (except 1). Also, all of them are guest accounts, no customer numbers.
- These are internal accounting write-offs, not customer transactions — a company marking a debt as uncollectible.
- We can drop these 6 rows. They're not sales, not returns of goods, not guest orders — they're just finance department bookkeeping.

In [31]:
print(df['Invoice'].str.len().value_counts())

Invoice
6    1013533
7      19104
Name: count, dtype: int64


In [32]:
df[(df['Invoice'].str.len() == 7) & (~df['Invoice'].str.startswith('C')) & (~df['Invoice'].str.startswith('A'))]

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,CustomerID,Country,InvoiceDateTime,InvoiceTime,InvoiceHour,InvoiceDayOfWeek,InvoiceMonth,is_guest


Note that now the only 7-letter Invoice numbers remaining are the Cancelled orders starting with a C.

Let's check the Quantity column to see any discrepancies.

In [33]:
invoice_prefix = df['Invoice'].str.extract(r'^([^\d]*)')[0]

for prefix in invoice_prefix.unique():
    subset = df[invoice_prefix == prefix]
    print(f"Prefix '{prefix}': {len(subset)} rows, Quantity range [{subset['Quantity'].min()}, {subset['Quantity'].max()}]")

Prefix '': 1013533 rows, Quantity range [-9600, 80995]
Prefix 'C': 19104 rows, Quantity range [-80995, 1]


We can see there are some invoices that are not cancelled (regular invoices) but still have a negative Quantity and there are Cancelled invoices with a positive quantity (of 1).  
Let's check them one by one.

In [34]:
df[(df['Invoice'].str.startswith('C')) & (df['Quantity'] > 0)]

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,CustomerID,Country,InvoiceDateTime,InvoiceTime,InvoiceHour,InvoiceDayOfWeek,InvoiceMonth,is_guest
76799,C496350,M,Manual,1,2010-02-01,373.57,GUEST_C496350,United Kingdom,2010-02-01 08:24:00,08:24:00,8,Monday,2010-02,True


In [35]:
df = df[~((df['Invoice'].str.startswith('C')) & (df['Quantity'] > 0))]

- Note that the StockCode is M and the description is Manual.
- This is used by the online retailer to manually key in miscellaneous charges, office adjustments, or custom fees onto an invoice.
- We will look into such codes further down in the notebook.
- For now, we can safely drop this one row as it represents non-product transactions, administrative fees, or manual ledger adjustments rather than physical inventory items.

Let's check the other case now where the quantity is negative for a regular invoice code. 

In [36]:
df[(df['Invoice'].str.len() == 6) & (df['Quantity'] < 0)]

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,CustomerID,Country,InvoiceDateTime,InvoiceTime,InvoiceHour,InvoiceDayOfWeek,InvoiceMonth,is_guest
263,489464,21733,RED HANGING HEART T-LIGHT HOLDER,-96,2009-12-01,0.0,GUEST_489464,United Kingdom,2009-12-01 10:52:00,10:52:00,10,Tuesday,2009-12,True
283,489463,71477,COLOUR GLASS. STAR T-LIGHT HOLDER,-240,2009-12-01,0.0,GUEST_489463,United Kingdom,2009-12-01 10:52:00,10:52:00,10,Tuesday,2009-12,True
284,489467,85123A,WHITE HANGING HEART T-LIGHT HOLDER,-192,2009-12-01,0.0,GUEST_489467,United Kingdom,2009-12-01 10:53:00,10:53:00,10,Tuesday,2009-12,True
3114,489655,20683,RAIN GIRL CHILDS UMBRELLA,-44,2009-12-01,0.0,GUEST_489655,United Kingdom,2009-12-01 17:26:00,17:26:00,17,Tuesday,2009-12,True
4296,489806,18010,ASSORTED FRAGRANCE BATH CONFETTI,-770,2009-12-02,0.0,GUEST_489806,United Kingdom,2009-12-02 12:42:00,12:42:00,12,Wednesday,2009-12,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1060794,581210,23395,BELLE JARDINIERE CUSHION COVER,-26,2011-12-07,0.0,GUEST_581210,United Kingdom,2011-12-07 18:36:00,18:36:00,18,Wednesday,2011-12,True
1060796,581212,22578,WOODEN STAR CHRISTMAS SCANDINAVIAN,-1050,2011-12-07,0.0,GUEST_581212,United Kingdom,2011-12-07 18:38:00,18:38:00,18,Wednesday,2011-12,True
1060797,581213,22576,SWALLOW WOODEN CHRISTMAS DECORATION,-30,2011-12-07,0.0,GUEST_581213,United Kingdom,2011-12-07 18:38:00,18:38:00,18,Wednesday,2011-12,True
1062371,581226,23090,VINTAGE GLASS T-LIGHT HOLDER,-338,2011-12-08,0.0,GUEST_581226,United Kingdom,2011-12-08 09:56:00,09:56:00,9,Thursday,2011-12,True


Notice how the Price for all of them seems to be 0 and the customer ID is a guest ID, so these are not saved customers. We can verify if it's true for all transactions.

In [37]:
# Verify if all such transactions are guest transactions
df1 = df[(df['Invoice'].str.len() == 6) & (df['Quantity'] < 0)]
df1[~df1['CustomerID'].str.startswith('GUEST')]

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,CustomerID,Country,InvoiceDateTime,InvoiceTime,InvoiceHour,InvoiceDayOfWeek,InvoiceMonth,is_guest


In [38]:
# Verify if the price is 0.0 for such transactions
df[(df['Invoice'].str.len() == 6) & (df['Quantity'] < 0) & (df['Price'] != 0.0)]

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,CustomerID,Country,InvoiceDateTime,InvoiceTime,InvoiceHour,InvoiceDayOfWeek,InvoiceMonth,is_guest


In [39]:
df['is_stock_adjustment'] = (
    (df['Quantity'] < 0) &
    (df['Price'] == 0) &
    (~df['Invoice'].str.startswith('C'))
)
print(df['is_stock_adjustment'].sum())

3067


In [40]:
df = df[~df['is_stock_adjustment']]
df.drop(['is_stock_adjustment'], axis=1, inplace=True)

- These 3067 rows appear to be internal stock-loss adjustments, identified by zero price, negative quantity, non-cancellation invoice, and 100% guest/unidentified customer. 
- These should be excluded from all customer-facing and revenue analysis as they're not purchases, not returns of purchases, and not attributable to any real customer.
- We can safely drop these rows.

We have seen above there are certain invoices starting with a 'C'. These are product cancellations or returns.  
We will create a is_cancellation flag to make it easier for us to analyse purchases v/s returns separately.

In [41]:
df['is_cancellation'] = df['Invoice'].str.startswith('C')

In [42]:
df.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,CustomerID,Country,InvoiceDateTime,InvoiceTime,InvoiceHour,InvoiceDayOfWeek,InvoiceMonth,is_guest,is_cancellation
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01,6.95,13085,United Kingdom,2009-12-01 07:45:00,07:45:00,7,Tuesday,2009-12,False,False
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01,6.75,13085,United Kingdom,2009-12-01 07:45:00,07:45:00,7,Tuesday,2009-12,False,False
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01,6.75,13085,United Kingdom,2009-12-01 07:45:00,07:45:00,7,Tuesday,2009-12,False,False
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01,2.10,13085,United Kingdom,2009-12-01 07:45:00,07:45:00,7,Tuesday,2009-12,False,False
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01,1.25,13085,United Kingdom,2009-12-01 07:45:00,07:45:00,7,Tuesday,2009-12,False,False


Let's move onto the StockCode column now. We can look at the different types of stock codes in our dataset.

In [43]:
# Real product codes are typically 5-digit numeric, optionally with 1-2 trailing letters
import re

def code_pattern_type(code):
    code = str(code)
    if re.match(r'^\d{5}[A-Za-z]{0,2}$', code):
        return 'standard_product'
    else:
        return 'non_standard'

df['stockcode_type'] = df['StockCode'].apply(code_pattern_type)
print(df['stockcode_type'].value_counts())

# Look at what the non-standard codes actually are
non_standard = df[df['stockcode_type'] == 'non_standard']
print(non_standard[['StockCode', 'Description']].value_counts())

stockcode_type
standard_product    1023630
non_standard           5939
Name: count, dtype: int64
StockCode     Description                        
POST          POSTAGE                                2086
DOT           DOTCOM POSTAGE                         1425
M             Manual                                 1386
C2            CARRIAGE                                277
D             Discount                                173
S             SAMPLES                                 101
BANK CHARGES  Bank Charges                            100
ADJUST        Adjustment by john on 26/01/2010 16      67
AMAZONFEE     AMAZON FEE                               36
DCGS0058      MISO PRETTY  GUM                         31
gift_0001_20  Dotcomgiftshop Gift Voucher £20.00       29
gift_0001_30  Dotcomgiftshop Gift Voucher £30.00       29
DCGSSGIRL     GIRLS PARTY BAG                          24
DCGSSBOY      BOYS PARTY BAG                           22
PADS          PADS TO MATCH ALL CUSHIONS 

- From the above results, it can be seen that some of the stock codes can be excluded completely. There is no real transaction value and they cannot contribute to any type of analysis.  
List of such stock codes: S (samples, given away — no real transaction value), BANK CHARGES, ADJUST/ADJUST2 (internal staff stock corrections), AMAZONFEE, CRUK (commission/fee), TEST001/TEST002 (literal test data), M, m (Manual Entries)

- Some of the codes can be used for revenue calculations, however should be excluded from product-level or basket-level analysis.
List of such stock codes: POST/DOT (Postage or Shipping Charges), C2 (Carriage), D (Discount), gift_000x (Gift vouchers).

- The remaining codes like DCGSxxxx or PADS can be left as products. They're genuine product codes flagged by the strict regex pattern. 

In [44]:
# Rows to drop entirely — not real, interpretable customer transactions
drop_codes = ['S', 'BANK CHARGES', ' BANK CHARGES', 'BANK CHARGES ', 'ADJUST', 'ADJUST ', 'ADJUST2',
              'AMAZONFEE', 'CRUK', 'TEST001', 'TEST002', 'M', 'm']
df = df[~df['StockCode'].isin(drop_codes)]

# Non-product codes: keep in revenue, exclude from product/basket analysis
non_product_revenue_codes = ['POST', 'DOT', 'C2', 'D'] + \
    [c for c in df['StockCode'].unique() if str(c).startswith('gift_')]
df['is_non_product'] = df['StockCode'].isin(non_product_revenue_codes)

df_products_only = df[~df['is_non_product']]   # for basket & product analysis
# df_clean itself remains your revenue-accurate, customer-level base

In [45]:
df.info()

<class 'pandas.DataFrame'>
Index: 1027838 entries, 0 to 1067370
Data columns (total 17 columns):
 #   Column            Non-Null Count    Dtype         
---  ------            --------------    -----         
 0   Invoice           1027838 non-null  str           
 1   StockCode         1027838 non-null  str           
 2   Description       1027838 non-null  str           
 3   Quantity          1027838 non-null  int64         
 4   InvoiceDate       1027838 non-null  object        
 5   Price             1027838 non-null  float64       
 6   CustomerID        1027838 non-null  object        
 7   Country           1027838 non-null  str           
 8   InvoiceDateTime   1027838 non-null  datetime64[us]
 9   InvoiceTime       1027838 non-null  object        
 10  InvoiceHour       1027838 non-null  int32         
 11  InvoiceDayOfWeek  1027838 non-null  str           
 12  InvoiceMonth      1027838 non-null  period[M]     
 13  is_guest          1027838 non-null  bool          
 14  is

### Add a revenue column

- Note that for all analysis, we're calculating the net revenue, so we will consider cancellations (returns) as well.
- This column does NOT represent gross sales or gross revenue.

In [46]:
df['Revenue'] = df['Quantity'] * df['Price']

In [47]:
df.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,CustomerID,Country,InvoiceDateTime,InvoiceTime,InvoiceHour,InvoiceDayOfWeek,InvoiceMonth,is_guest,is_cancellation,stockcode_type,is_non_product,Revenue
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01,6.95,13085,United Kingdom,2009-12-01 07:45:00,07:45:00,7,Tuesday,2009-12,False,False,standard_product,False,83.4
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01,6.75,13085,United Kingdom,2009-12-01 07:45:00,07:45:00,7,Tuesday,2009-12,False,False,standard_product,False,81.0
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01,6.75,13085,United Kingdom,2009-12-01 07:45:00,07:45:00,7,Tuesday,2009-12,False,False,standard_product,False,81.0
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01,2.10,13085,United Kingdom,2009-12-01 07:45:00,07:45:00,7,Tuesday,2009-12,False,False,standard_product,False,100.8
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01,1.25,13085,United Kingdom,2009-12-01 07:45:00,07:45:00,7,Tuesday,2009-12,False,False,standard_product,False,30.0


### Save the Cleaned Dataset

In [48]:
df.info()

<class 'pandas.DataFrame'>
Index: 1027838 entries, 0 to 1067370
Data columns (total 18 columns):
 #   Column            Non-Null Count    Dtype         
---  ------            --------------    -----         
 0   Invoice           1027838 non-null  str           
 1   StockCode         1027838 non-null  str           
 2   Description       1027838 non-null  str           
 3   Quantity          1027838 non-null  int64         
 4   InvoiceDate       1027838 non-null  object        
 5   Price             1027838 non-null  float64       
 6   CustomerID        1027838 non-null  object        
 7   Country           1027838 non-null  str           
 8   InvoiceDateTime   1027838 non-null  datetime64[us]
 9   InvoiceTime       1027838 non-null  object        
 10  InvoiceHour       1027838 non-null  int32         
 11  InvoiceDayOfWeek  1027838 non-null  str           
 12  InvoiceMonth      1027838 non-null  period[M]     
 13  is_guest          1027838 non-null  bool          
 14  is

In [49]:
# Check CustomerID
print(df['CustomerID'].apply(type).value_counts())

CustomerID
<class 'int'>    796667
<class 'str'>    231171
Name: count, dtype: int64


In [50]:
# Fix customer id before saving
df['CustomerID'] = df['CustomerID'].astype(str)

# Drop this and load again everytime as Period datatype will not be carried forward while saving the file.
df = df.drop(columns=['InvoiceMonth'])

In [51]:
df.info()

<class 'pandas.DataFrame'>
Index: 1027838 entries, 0 to 1067370
Data columns (total 17 columns):
 #   Column            Non-Null Count    Dtype         
---  ------            --------------    -----         
 0   Invoice           1027838 non-null  str           
 1   StockCode         1027838 non-null  str           
 2   Description       1027838 non-null  str           
 3   Quantity          1027838 non-null  int64         
 4   InvoiceDate       1027838 non-null  object        
 5   Price             1027838 non-null  float64       
 6   CustomerID        1027838 non-null  str           
 7   Country           1027838 non-null  str           
 8   InvoiceDateTime   1027838 non-null  datetime64[us]
 9   InvoiceTime       1027838 non-null  object        
 10  InvoiceHour       1027838 non-null  int32         
 11  InvoiceDayOfWeek  1027838 non-null  str           
 12  is_guest          1027838 non-null  bool          
 13  is_cancellation   1027838 non-null  bool          
 14  st

In [52]:
# Save the cleaned dataset
df.to_parquet('data/cleaned_data.parquet', engine='pyarrow', index=False)

## Summary

- The raw Online Retail II data was consolidated and cleaned to create a consistent transaction-level dataset for subsequent customer and product analytics.
- Duplicate records, invalid transactions, stock adjustments and unusable product records were identified and handled based on their business meaning.
- Missing customer identifiers were retained and flagged as guest transactions, allowing them to remain available for transaction-level and revenue analysis while excluding them from analyses requiring persistent customer identities.
- Cancellation and non-product transactions were also identified through dedicated flags to allow them to be included or excluded appropriately in later analyses.

The resulting dataset provides the foundation for the subsequent business performance, customer, product, retention, CLV and market basket analyses.